### PR TITLE
Refactor sample validation

### PR DESCRIPTION
**Pull request by NickGrantONS**
# Motivation and Context
The 'validation' of the samples was spaghetti code and highly inefficient.

Given the need to load large samples for ASHE etc. it seemed prudent to refactor all the code that the lies on the sample load pipeline, to make it more performant and reliable.

In the course of the refactoring, a number of bugs were spotted, such as inability to recover from momentary connection problems with the Survey, Collection Instrument and Party services - these should retry without manual intervention.

The code was inefficient, badly performant, hard to understand, maintain, and it did not handle errors well.

# What has changed
Major refactoring of the `ValidateSampleUnits` class.

Sadly it's still not quite honouring the single responsibility principle, but the code is a lot less complex and should make future refactoring easier.

Substantial improvement to performance and error handling.

# How to test?
Load a big sample. Make sure the sample validates OK and the collection exercise changes to "Ready for Live" status.

# Links
Trello: https://trello.com/c/eAmhl3iB/729-use-stream-and-multithreading-in-validatesampleunits-to-rapidly-validate-the-samples-13

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java [hunk 1]
```diff
@@ -68,75 +58,110 @@
...
+            try {
+              sampleUnitGroup.setStateFK(
+                  sampleUnitGroupState.transition(
+                      sampleUnit.getSampleUnitGroup().getStateFK(), event));
             } catch (CTPException e) {
-              log.error(
-                  "Error validating collection exercise, collectionExerciseId: {}",
-                  exercise.getId());
-              log.error("Stack trace: {}", e);
+              // Ignored because this should never happen - we know the transition is valid
```
Comment (necrophonic) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java]: Is it safe to ever do this? Shouldn't it at least log? _Things that should never happen_ tend to happen more often than you think :)

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidationScheduler.java [hunk 2]
```diff
@@ -11,17 +14,34 @@
...
+    try {
+      if (lock.tryLock(1, TimeUnit.HOURS)) {
+        try {
+          sampleService.validateSampleUnits();
+        } finally {
+          lock.unlock();
+        }
+      }
+    } catch (InterruptedException e) {
+      // Ignored
```
Comment (necrophonic) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidationScheduler.java]: Should probably at least log

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/repository/SampleUnitRepository.java [hunk 3]
```diff
@@ -51,4 +53,8 @@ boolean existsBySampleUnitRefAndSampleUnitTypeAndSampleUnitGroupCollectionExerci
    */
   List<ExerciseSampleUnit> findBySampleUnitGroupCollectionExerciseAndSampleUnitGroupStateFK(
       CollectionExercise collectionExercise, SampleUnitGroupState sampleUnitGroupState);
+
```
Comment (mfcrocker) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/repository/SampleUnitRepository.java]: It'd be nice to have Javadoc

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java [hunk 4]
```diff
@@ -45,19 +42,12 @@
 /** Class responsible for business logic to validate SampleUnits. */
 @Component
 public class ValidateSampleUnits {
+
```
Comment (mfcrocker) [inline src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java]: :art:

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java (hunk 1 above)
Comment (mfcrocker) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java]: At the very least it'll mean it'll start screaming if we mess with the state machine and introduce a regression

Comment (NickGrantONS) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java]: it's the state machine code that's wrong - it shouldn't throw a checked exception. I suppose I could re-throw the exception as a runtime exception

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidationScheduler.java (hunk 2 above)
Comment (NickGrantONS) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidationScheduler.java]: this happens only in the rare event that the thread is being killed by a shutdown. I'm not sure we'd have enough time to log and what would we log except to say that the thread was forcibly killed, which isn't really an error or anything.

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/repository/SampleUnitRepository.java (hunk 3 above)
Comment (NickGrantONS) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/repository/SampleUnitRepository.java]: done

**Comment by NickGrantONS** [comment]
Am investigating some potential problem this has introduced with the Quartz scheduler and collexes flipping to live automatically.

Got this stack trace:
`2019-03-12 15:44:47.011 ERROR [ONS CollectionExerciseSvc,e147675438a75689,ba6209f8143aa881,false]  6 --- [ean)#37fc7e3c-1] m.CollectionExerciseEventInboundReceiver : Failed to set collection exercise to LIVE state; collection_exercise_id="2f144595-2dc9-4589-8a1c-0e41e4bfa589"

uk.gov.ons.ctp.common.error.CTPException: State Transition from READY_FOR_REVIEW via GO_LIVE is forbidden.
	at uk.gov.ons.ctp.common.state.BasicStateTransitionManager.transition(BasicStateTransitionManager.java:47)
	at uk.gov.ons.ctp.response.collection.exercise.service.CollectionExerciseService.transitionCollectionExercise(CollectionExerciseService.java:730)
	at uk.gov.ons.ctp.response.collection.exercise.service.CollectionExerciseService.transitionCollectionExercise(CollectionExerciseService.java:760)
	at uk.gov.ons.ctp.response.collection.exercise.service.CollectionExerciseService$$FastClassBySpringCGLIB$$758f9914.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:669)
	at uk.gov.ons.ctp.response.collection.exercise.service.CollectionExerciseService$$EnhancerBySpringCGLIB$$9f62f43d.transitionCollectionExercise(<generated>)
	at uk.gov.ons.ctp.response.collection.exercise.message.CollectionExerciseEventInboundReceiver.transitionCollectionExerciseToLive(CollectionExerciseEventInboundReceiver.java:81)
	at uk.gov.ons.ctp.response.collection.exercise.message.CollectionExerciseEventInboundReceiver.handleEventMessage(CollectionExerciseEventInboundReceiver.java:49)
	at sun.reflect.GeneratedMethodAccessor377.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.expression.spel.support.ReflectiveMethodExecutor.execute(ReflectiveMethodExecutor.java:113)
	at org.springframework.expression.spel.ast.MethodReference.getValueInternal(MethodReference.java:102)
	at org.springframework.expression.spel.ast.MethodReference.access$000(MethodReference.java:49)
	at org.springframework.expression.spel.ast.MethodReference$MethodValueRef.getValue(MethodReference.java:347)
	at org.springframework.expression.spel.ast.CompoundExpression.getValueInternal(CompoundExpression.java:88)
	at org.springframework.expression.spel.ast.SpelNodeImpl.getTypedValue(SpelNodeImpl.java:131)
	at org.springframework.expression.spel.standard.SpelExpression.getValue(SpelExpression.java:360)
	at org.springframework.integration.util.AbstractExpressionEvaluator.evaluateExpression(AbstractExpressionEvaluator.java:169)
	at org.springframework.integration.util.MessagingMethodInvokerHelper.processInternal(MessagingMethodInvokerHelper.java:319)
	at org.springframework.integration.util.MessagingMethodInvokerHelper.process(MessagingMethodInvokerHelper.java:155)
	at org.springframework.integration.handler.MethodInvokingMessageProcessor.processMessage(MethodInvokingMessageProcessor.java:93)
	at org.springframework.integration.handler.ServiceActivatingHandler.handleRequestMessage(ServiceActivatingHandler.java:89)
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:109)
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:127)
	at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:116)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:148)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:121)
	at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:89)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:423)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:373)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:115)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:45)
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:105)
	at org.springframework.integration.handler.AbstractMessageProducingHandler.sendOutput(AbstractMessageProducingHandler.java:360)
	at org.springframework.integration.handler.AbstractMessageProducingHandler.produceOutput(AbstractMessageProducingHandler.java:271)
	at org.springframework.integration.handler.AbstractMessageProducingHandler.sendOutputs(AbstractMessageProducingHandler.java:188)
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:115)
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:127)
	at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:116)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:148)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:121)
	at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:89)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:423)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:373)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:115)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:45)
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:105)
	at org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:188)
	at org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter.access$1100(AmqpInboundChannelAdapter.java:56)
	at org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter$Listener.processMessage(AmqpInboundChannelAdapter.java:246)
	at org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter$Listener.onMessage(AmqpInboundChannelAdapter.java:203)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:822)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.invokeListener(AbstractMessageListenerContainer.java:745)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.access$001(SimpleMessageListenerContainer.java:97)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$1.invokeListener(SimpleMessageListenerContainer.java:189)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.invokeListener(SimpleMessageListenerContainer.java:1276)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.executeListener(AbstractMessageListenerContainer.java:726)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.doReceiveAndExecute(SimpleMessageListenerContainer.java:1219)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.receiveAndExecute(SimpleMessageListenerContainer.java:1189)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.access$1500(SimpleMessageListenerContainer.java:97)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.run(SimpleMessageListenerContainer.java:1421)
	at java.lang.Thread.run(Thread.java:748)
`

File: src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java (hunk 1 above)
Comment (necrophonic) [reply src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java]: We just don't want exceptions disappearing into the ether if they occur (even if they're really not supposed to). Everything should be handled even if that's just "something that shouldn't have happened has happened".

**Pull request merged**